### PR TITLE
Add HTTP headers support to common HTTP client.

### PR DESCRIPTION
**Pull request by roidelapluie**
This is named `http_headers` so it does not clash with blackbox exporter's headers and Prometheus remote client's headers, which are simple maps.

Signed-off-by: Julien Pivotto <roidelapluie@o11y.eu>

### CODE REVIEW

**Comment by roidelapluie** [comment]
Mailing list discussion: https://groups.google.com/g/prometheus-developers/c/p9HTlPzwUEo

File: config/headers.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			_, err := os.ReadFile(f)
+			if err != nil {
+				return fmt.Errorf("unable to read header %q from file %s: %w", http.CanonicalHeaderKey(k), f, err)
+			}
+		}
+	}
+	for k := range uniqueHeaders {
+		if _, ok := reservedHeaders[http.CanonicalHeaderKey(k)]; ok {
+			return fmt.Errorf("setting header %q is not allowed", http.CanonicalHeaderKey(k))
+		}
```
Comment (LeviHarrison) [inline config/headers.go:96]: It would make more sense for this check to come first

**Comment by zetaab** [comment]
@roidelapluie @LeviHarrison what is status in this PR? At least I am interested of this feature

**Comment by Dronec** [comment]
Me too! Adding headers to requests is a relatively trivial operation and multiple APIs support that

**Comment by LeviHarrison** [comment]
Just some small revisions

@roidelapluie friendly ping

**Review by LeviHarrison — APPROVED** [review]
LGTM, thanks

As usual we'll need to update Prom docs after this

**Comment by pracucci** [comment]
@roidelapluie I think this PR introduced an issue in JSON marshalling. Could you take a look at https://github.com/prometheus/common/pull/651 please?

File: config/http_config.go [hunk 2]
```diff
@@ -311,6 +311,9 @@ type HTTPClientConfig struct {
 	EnableHTTP2 bool `yaml:"enable_http2" json:"enable_http2"`
 	// Proxy configuration.
 	ProxyConfig `yaml:",inline"`
+	// HTTPHeaders specify headers to inject in the requests. Those headers
+	// could be marshalled back to the users.
+	HTTPHeaders *Headers `yaml:"http_headers" json:"http_headers"`
```
Comment (yeya24) [inline config/http_config.go:316]: Should we at least make it omit empty?

**Comment by udf2457** [comment]
@roidelapluie any chance you could update the docs to reflect the existence of `http_headers` ?

CTRL-F of [the docs](https://prometheus.io/docs/prometheus/latest/configuration/configuration/) yields zero results for `http_headers`

**Pull request merged**